inspirehep/dojson/hep/fields/bd01x09x.py

```diff
@@ -59,29 +59,56 @@ def dois(self, key, value):
             if isinstance(val.get("2"), list):
                 if val.get("2", '')[0].lower() == "doi":
                     if isinstance(val.get('a'), list):
-                        for v in val.get('a'):
+                        if isinstance(val.get('9'), list):
```

Comment:
As DOIs are quite sensitive, we need some test cases here. You can add some different MarcXML snippets with just the doi field with cases you have found problematic.

Then call:

```python
from dojson.contrib.marc21.utils import create_record
x = create_record(snippet)
hep.do(x)
```

And make sure that the outputted DOIs are clean and no data is lost.